File: desktop/src/utils/platform.ts
import { invoke } from "@tauri-apps/api/core";

export interface PlatformInfo {
  os: string;
  voiceHotkey: string;
  toolbarHotkey: string;
  pasteShortcut: string;
  modKey: string;
}

function fallbackPlatformInfo(): PlatformInfo {
  const isMac = navigator.userAgent.includes("Mac");
  return {
    os: isMac ? "macos" : "windows",
    voiceHotkey: "Ctrl+Space",
    toolbarHotkey: "Ctrl+Space×2",
    pasteShortcut: isMac ? "Cmd+V" : "Ctrl+V",
    modKey: isMac ? "Cmd" : "Ctrl",
  };
}

let cached: PlatformInfo | null = null;

export async function getPlatformInfo(): Promise<PlatformInfo> {
  if (cached) return cached;
  try {
    cached = await invoke<PlatformInfo>("get_platform_info");
  } catch {
    cached = fallbackPlatformInfo();
  }
  return cached;
}

export function getPlatformInfoSync(): PlatformInfo {
  return cached ?? fallbackPlatformInfo();
}

export function isMacOS(): boolean {
  return getPlatformInfoSync().os === "macos";
}

File: desktop/src-tauri/src/platform.rs
use serde::Serialize;

#[derive(Serialize)]
#[serde(rename_all = "camelCase")]
pub struct PlatformInfo {
    pub os: &'static str,
    pub voice_hotkey: &'static str,
    pub toolbar_hotkey: &'static str,
    pub paste_shortcut: &'static str,
    pub mod_key: &'static str,
}

pub fn current_os() -> &'static str {
    if cfg!(target_os = "macos") {
        "macos"
    } else if cfg!(target_os = "windows") {
        "windows"
    } else if cfg!(target_os = "linux") {
        "linux"
    } else {
        "other"
    }
}

pub fn platform_info() -> PlatformInfo {
    if cfg!(target_os = "macos") {
        PlatformInfo {
            os: "macos",
            voice_hotkey: "Ctrl+Space",
            toolbar_hotkey: "Ctrl+Space×2",
            paste_shortcut: "Cmd+V",
            mod_key: "Cmd",
        }
    } else {
        PlatformInfo {
            os: current_os(),
            voice_hotkey: "Ctrl+Space",
            toolbar_hotkey: "Ctrl+Space×2",
            paste_shortcut: "Ctrl+V",
            mod_key: "Ctrl",
        }
    }
}

#[tauri::command]
pub fn get_platform_info() -> PlatformInfo {
    platform_info()
}

pub fn voice_hotkey_label() -> &'static str {
    "Ctrl+Space"
}

pub fn tray_tooltip() -> String {
    format!("DeltaVoice - {} to record", voice_hotkey_label())
}
